Sql template

vietj opened the pull request:
SQL templates for Vert.x 4

Latest generated documentation https://gist.github.com/vietj/db18ad21ea242f7260a8a824f6d6d3a8

BillyYccc commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/BatchTemplate.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  @GenIgnore
+  void batch(List<T> list, Handler<AsyncResult<Void>> result);
```
How about naming it `batchUpdate` since it returns an asynchronous result of `Void` type?

knotenpunkt commented:
when do you think to merge this PR?^^

And Julien, you should really take a look on my two BUG-issues in the vertx-repository.
The other once, are more or less feature request, wrong labeled, you should adjust them.

vietj commented:
link to current doc https://gist.github.com/vietj/db18ad21ea242f7260a8a824f6d6d3a8

zenios commented:
Hi

My only concern is with annotation pollution. To many annotations needs to be added to a class in order to be @RowMapped and @ParamsMapped.

Why don't we have @TemplateParam and @Column added to all getters by default and provide a way to exclude it. Just like lombok does for getter / setter generation

In the end most people map one to one from table to bean

vietj commented:
@zenios this is a good concern and I believe it is a possible choice given we provide a way to exclude some fields from the mapping. Keep in mind also that such data object can also be mapped to/from JSON using the json code generator too

vietj commented:
alright folks, this is now ready for review

jponge commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,325 @@
+[source,groovy]
+----
+dependencies {
+  compile '${maven.groupId}:${maven.artifactId}:${maven.version}'
```
Use `implementation` rather than `compile`.

jponge replied:
But that's a more general remark. `compile` is still here, but it's recommended to go with `implementation` for dependencies you consume, and `api` for dependencies you want to consume the re-export.

jponge commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,325 @@
+{@link examples.TemplateExamples#queryExample}
+----
+
+When you need to perform an insert or update operation and you do not care of the result, you can use instead {@link io.vertx.sqlclient.template.SqlTemplate#forUpdate}:
```
you can use `{@link XYZ}` instead.

jponge commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,325 @@
+
+== Row mapping
+
+Templates produces {@link io.vertx.sqlclient.Row} as default result.
```
By default templates produce `{...}` as result type.

jponge commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,325 @@
+
+Templates produces {@link io.vertx.sqlclient.Row} as default result.
+
+You can provide a function:
```
I'd put `to achieve row level mapping instead:` here with the full function and template call below

jponge commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,325 @@
+
+[source]
+----
+compile "io.vertx:vertx-codegen:${maven.version}:processor"
```
`annotationProcessor` scope in Gradle

jponge commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,325 @@
+compile "io.vertx:vertx-sql-client-template:${maven.version}"
+----
+
+IDE provides usually support for annotation processors.
```
usually provide

jponge commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/SqlTemplate.java:
```diff
@@ -0,0 +1,145 @@
+package io.vertx.sqlclient.template;
```
BTW you should have 2020 license headers in files (IntelliJ copyright update can help)

zenios commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/impl/SqlTemplate.java:
```diff
@@ -0,0 +1,50 @@
+
+public class SqlTemplate {
+
+  private static Pattern PARAM_PATTERN = Pattern.compile(":(\\p{javaUnicodeIdentifierStart}\\p{javaUnicodeIdentifierPart}*)");
```
This will not work for postgresql :: cast

This is how spring does it
https://github.com/spring-projects/spring-framework/blob/4ca64684f6d5b7414ca3628a6b372a7c586fd2a6/spring-jdbc/src/main/java/org/springframework/jdbc/core/namedparam/NamedParameterUtils.java#L82

Another solution is to use ${name} via this pattern

```
Pattern.compile("\\$\\{\\s*(.*?)\\s*}");
```

We could also cache the result

vietj replied:
I like this idea

zenios commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/impl/SqlTemplateImpl.java:
```diff
@@ -0,0 +1,108 @@
+    return sqlTemplate.mapTuple(paramsMapper.apply(params));
+  }
+
+  @Override
+  public void execute(T parameters, Handler<AsyncResult<R>> handler) {
+
+    queryMapper
+      .apply(client.preparedQuery(sqlTemplate.getSql()))
+      .execute(toTuple(parameters), handler);
+  }
```
This can be remove if we use the default method on interface

vietj replied:
This is important to keep this and delegate to the original query

zenios commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/impl/SqlTemplateImpl.java:
```diff
@@ -0,0 +1,108 @@
+      .execute(toTuple(params));
+  }
+
+  @Override
+  public void executeBatch(List<T> batch, Handler<AsyncResult<R>> handler) {
+    queryMapper.apply(client.preparedQuery(sqlTemplate.getSql()))
+      .executeBatch(batch
+        .stream()
+        .map(this::toTuple)
+        .collect(Collectors.toList()), handler);
+  }
```
This can be remove if we use the default method on interface

vietj replied:
This is important to keep this and delegate to the original query

zenios commented on vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlClientInternal.java:
```diff
@@ -0,0 +1,17 @@
+public interface SqlClientInternal extends SqlClient {
+
+  /**
+   * Append a query place holder in the {@code query}.
```
placeholder not place holder

vietj commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/SqlTemplate.java:
```diff
@@ -0,0 +1,145 @@
+   * @param batch the batch
+   * @param handler the result handler
+   */
+  void executeBatch(List<I> batch, Handler<AsyncResult<R>> handler);
```
because in practice the underlying query will actually use internally a `Promise<R>` that is created from the handler, so if you pass a promise to an `executeBatch` it will not be rewrapped into another promise.

vietj commented on vertx-sql-client-template/src/main/java/examples/TemplateExamples.java:
```diff
@@ -0,0 +1,397 @@
+    return params;
+  };
+
+  public void paramsUserMapper() {
+    Function<User, Map<String, Object>> PARAMS_USER_MAPPER = user -> {
+      Map<String, Object> params = new HashMap<>();
+      params.put("id", user.id);
+      params.put("firstName", user.firstName);
+      params.put("lastName", user.lastName);
+      return params;
+    };
+  }
```
this is referenced from `index.adoc`

vietj commented on vertx-sql-client-template/src/main/java/examples/TemplateExamples.java:
```diff
@@ -0,0 +1,397 @@
+    return user;
+  };
+
+  public void rowUserMapper() {
+    Function<Row, User> ROW_USER_MAPPER = row -> {
+      User user = new User();
+      user.id = row.getInteger("id");
+      user.firstName = row.getString("firstName");
+      user.lastName = row.getString("lastName");
+      return user;
+    };
+  }
```
this is referenced from `index.adoc`

vietj commented:
I would prefer to keep `@Column` and `@TemplateParam` distincts because they don't have the same meaning.

vietj commented:
@zenios @jponge addressed comments - now the syntax uses `${...}` because using `:` syntax would require to write an SQL parser which is a non goal

zenios commented:
Will do another later.thanks

On Thu, 7 May 2020, 17:08 Julien Viet, <notifications@github.com> wrote:

> @zenios <https://github.com/zenios> @jponge <https://github.com/jponge>
> addressed comments - now the syntax uses ${...} because using : syntax
> would require to write an SQL parser which is a non goal
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-vertx/vertx-sql-client/pull/538#issuecomment-625277932>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAV3YP3GS5FZCZXWJKL64DRQK6E5ANCNFSM4LHBLQQQ>
> .
>

vietj commented:
Thank you.

> On 7 May 2020, at 16:09, zenios <notifications@github.com> wrote:
> 
> 
> Will do another later.thanks
> 
> On Thu, 7 May 2020, 17:08 Julien Viet, <notifications@github.com> wrote:
> 
> > @zenios <https://github.com/zenios> @jponge <https://github.com/jponge>
> > addressed comments - now the syntax uses ${...} because using : syntax
> > would require to write an SQL parser which is a non goal
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/eclipse-vertx/vertx-sql-client/pull/538#issuecomment-625277932>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AAAV3YP3GS5FZCZXWJKL64DRQK6E5ANCNFSM4LHBLQQQ>
> > .
> >
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse-vertx/vertx-sql-client/pull/538#issuecomment-625278840>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABXDCUH32IS7DC77I7ZNRLRQK6LDANCNFSM4LHBLQQQ>.
>

zenios commented on vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlClientInternal.java:
```diff
@@ -5,13 +5,13 @@
+   * @param index the placeholder index
    */
-  int appendQueryPlaceHolder(StringBuilder queryBuilder, int index, int current);
+  int appendQueryPlaceholder(StringBuilder queryBuilder, int index, int current);
```
We should document the return value

zenios commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,332 @@
+= Client SQL templates
```
SQL Client templates

vietj replied:
we might actually rename this to simply "SQL templates"

zenios commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,332 @@
+= Client SQL templates
+:toc:
+
+SQL client templates is a small library design to facilitate the execution of SQL queries.
```
library designed

zenios commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,332 @@
+
+== Usage
+
+To use client SQL templates add the following dependency to the _dependencies_ section of your build descriptor:
```
To use SQL client templates

zenios commented on vertx-sql-client-template/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,332 @@
+compile "io.vertx:vertx-sql-client-template:${maven.version}"
+----
+
+IDE usually provide usually support for annotation processors.
```
IDEs usually provide support for annotation processors.

zenios commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/SqlTemplate.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.stream.Collector;
+
+/**
+ * A SQL template.
```
An SQL template.

zenios commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/SqlTemplate.java:
```diff
@@ -0,0 +1,145 @@
+public interface SqlTemplate<I, R> {
+
+  /**
+   * Create a SQL template for query purpose consuming map parameters and returning {@link Row}.
```
Create an SQL template

zenios commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/SqlTemplate.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  /**
+   * Create a SQL template for query purpose consuming map parameters and returning void.
```
Create an SQL template

BillyYccc commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/impl/SqlTemplateImpl.java:
```diff
@@ -0,0 +1,108 @@
+
+  public SqlTemplateImpl(SqlClient client,
+                         io.vertx.sqlclient.template.impl.SqlTemplate sqlTemplate,
+                         Function<PreparedQuery<RowSet<Row>>, PreparedQuery<R>> foobar,
```
`foobar` should be named `queryMapper`?

BillyYccc commented on vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlClientBase.java:
```diff
@@ -37,7 +37,13 @@
+
+  @Override
+  public int appendQueryPlaceholder(StringBuilder queryBuilder, int index, int current) {
+    queryBuilder.append("?");
```
we need to override this method in `MSSQLConnectionImpl` and `MSSQLPoolImpl` for sql server client.
The syntax is like postgres as:
```
    queryBuilder.append('@').append(1 + index);
    return index;
```

vietj replied:
oh good to know I will do it

BillyYccc commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/generator/RowMapperGen.java:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  private static String getter(TypeInfo type) {
+    switch (type.getKind()) {
```
it seems we don't support mapping string/numeric datat type to enum type(name/ordinal). I wonder if we could support this here.

vietj replied:
I think we can map it to get/set String easily

vietj replied:
at least for beginning, some might want to provide at some point a custom function for this type

vietj replied:
I've checked in such mapping support in master for PostgreSQL and implement support for templates in this PR.

BillyYccc commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/generator/RowMapperGen.java:
```diff
@@ -0,0 +1,215 @@
+        case "io.vertx.core.buffer.Buffer":
+          return "getBuffer";
+      }
+      DataObjectInfo dataObject = type.getDataObject();
```
can we add tests for data object type like geometrics data object and numeric data object to check whether it works?

vietj replied:
yes, I'm not sure they work, we'll write the tests and figure out after :-)

vietj replied:
I added support and tests for such types

vietj replied:
Actually they are mapped to generic types with `row.get(Box.class, pos)`

BillyYccc commented on vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/generator/RowMapperGen.java:
```diff
@@ -0,0 +1,215 @@
+    }
+    if (type instanceof ClassTypeInfo) {
+      ClassTypeInfo ct = (ClassTypeInfo) type;
+      switch (ct.getName()) {
```
MySQL is using `java.time.Duration` type, can we add the mapping here using Row#get(java.time.Duration.class, index)?

vietj replied:
I will try adding a specific tests for this using test container and implement it.

vietj replied:
I added support for this and specific integration tests

vietj commented:
Ready for another review

vietj commented:
Gist with docs has also be updated https://gist.github.com/vietj/db18ad21ea242f7260a8a824f6d6d3a8

vietj commented:
Any more feedback on this ?

zenios commented:
@vietj Skipping of fields will not be included on version 1?

BillyYccc commented on vertx-sql-client-templates/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,338 @@
+The _codegen_ annotation processor generates these classes at compilation time. It is a feature of the Java
+compiler so _no extra step_ is required, it is just a matter of configuring correctly your build:
+
+Just add the `io.vertx:vertx-codegen:processor` and `io.vertx:vertx-sql-client-template`
```
should be `vertx-sql-client-templates`?

vietj replied:
right

BillyYccc commented on vertx-sql-client-templates/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,338 @@
+</dependency>
+<dependency>
+  <groupId>io.vertx</groupId>
+  <artifactId>vertx-sql-client-template</artifactId>
```
should be `vertx-sql-client-templates`?

vietj replied:
right

BillyYccc commented on vertx-sql-client-templates/src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,338 @@
+[source]
+----
+annotationProcessor "io.vertx:vertx-codegen:${maven.version}:processor"
+compile "io.vertx:vertx-sql-client-template:${maven.version}"
```
`vertx-sql-client-templates`?

vietj replied:
right
